### PR TITLE
Highlights: anchored dialogs

### PR DESCRIPTION
**Pull request by hius07**
We now have cached screen boxes, let's use them to calculate the position of anchored popup dialogs.
Applicable to Dictionary, Highlight, Edit highlight, Note dialog windows.
Also fixes difficult "non-page view" cases when anchored dialog position was reset to "center".

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12810)

### CODE REVIEW

File: frontend/document/credocument.lua [hunk 1]
```diff
@@ -714,12 +714,12 @@ function CreDocument:getTextFromPositions(pos0, pos1, do_not_draw_selection)
         drawSelection, drawSegmentedSelection)
     logger.dbg("CreDocument: get text range", text_range)
     if text_range then
-        -- local line_boxes = self:getScreenBoxesFromPositions(text_range.pos0, text_range.pos1)
+        local line_boxes = self:getScreenBoxesFromPositions(text_range.pos0, text_range.pos1)
```
Comment (poire-z) [inline frontend/document/credocument.lua:717]: getScreenBoxesFromPositions() may be expensive, wondering if there are other calls to getTextFromPositions() that would not need it, and then if calling it and returning sboxes may be a parameter to that getTextFromPositions().

**Review by poire-z — COMMENTED** [review]
(Trusting you on the box x/y/h maths :))

File: frontend/document/credocument.lua (hunk 1 above)
Comment (hius07) [reply frontend/document/credocument.lua:717]: Yes, there are only 2 callers that do not require it:
https://github.com/koreader/koreader/blob/5d68ece781519245850b8d011a56de109f70738b/frontend/apps/reader/modules/readerhighlight.lua#L1847
https://github.com/koreader/koreader/blob/5d68ece781519245850b8d011a56de109f70738b/frontend/apps/reader/modules/readerview.lua#L1407
I thought the parameter was not needed.

**Comment by poire-z** [comment]
Just a note, I don't know if this is an issue or not.
In this PR, you un-commented a line that was commented out since 2016 in credocument.lua:
`local line_boxes = self:getScreenBoxesFromPositions(text_range.pos0, text_range.pos1)`
and you added one that uses that same function:
`new_pboxes = self.document:getScreenBoxesFromPositions(new_pos0, new_pos1)`.

When selecting text in a LTR, we get a box per line of selected text. But with some RTL content, we get a box per character:
image

In 2018, I added a 3rd paremeter to this getScreenBoxesFromPositions(pos0, pos1, `get_segments`), which I think I have set to `true` in all our other calls of this function.
And when we do, we get a small number of rects in sbox:
image

I'm not familiar with this PR and what you do with these boxes, but you might want to check how it behave with RTL text. If there's something odd, may be adding the `true` can solve the issue.
(It may be more work in C when we use get_segments, but it would then be less boxes to iterate on on the Lua side :))
(I'm not forcing that on you, but having these use `true` like everything else would probably better for consistency - again not familiar with these PR to test if using `true` have it still work as expected.)

**Comment by poire-z** [comment]
> We now have cached screen boxes, let's use them to calculate the position of anchored popup dialogs.
> Applicable to Dictionary, Highlight, Edit highlight, Note dialog windows.

@hius07 : looks like I always get these in the middle of the screen (on EPUB, regular LTR book).
Is this still working for you ? Or did I misundertand the feature ?

image
image
image

**Comment by hius07** [comment]
It works.
It follows the setting Long-press on text - Highlight dialog position.

**Pull request merged**